lib/src/device.dart

```
@@ -232,6 +236,100 @@ class AndroidDevice extends _Device {
+    } catch (e) {
+      // nc returns a non-zero exit code when the port is _not_ in use,
+      // so we can start the server if we catch an error.
+      await Process.start('pub', ['run', 'sky_tools:sky_server', _serverPort],
+          workingDirectory: serverRoot, mode: ProcessStartMode.DETACHED);
+    }
+
+    // Set up reverse port-forwarding so that the Android app can reach the
+    // server running on localhost.
+    String serverPortString = 'tcp:$_serverPort';
+    runCheckedSync([adbPath, 'reverse', serverPortString, serverPortString]);
+
+    String relativeDartMain = path.relative(mainDart, from: serverRoot);
+    String url = 'http://localhost:$_serverPort/$relativeDartMain';
+    if (poke) {
+      url += '?rand=${new Random().nextDouble()}';
+    }
+
+    // Actually launch the app on Android
+    List<String> cmd = [
+      adbPath,
+      'shell',
+      'am',
+      'start',
+      '-a',
+      'android.intent.action.VIEW',
+      '-d',
+      url,
+    ];
+
+    if (checked) {
+      cmd.addAll(['--ez', 'enable-checked-mode', 'true']);
+    }
+
+    cmd.add(apk.component);
+    runCheckedSync(cmd);
+
+    return true;
+  }
+
+  bool stop(AndroidApk apk) {
+    // Turn off reverse port forwarding
+    try {
+      runCheckedSync([adbPath, 'reverse', '--remove', 'tcp:$_serverPort']);
+    } catch (e) {}
+    // Stop the app
+    runCheckedSync([adbPath, 'shell', 'am', 'force-stop', apk.appPackageID]);
+    // Kill the server
+    try {
+      if (Platform.isMacOS) {
+        String pid = runCheckedSync(['lsof', '-i', ':$_serverPort', '-t']);
+        // Killing a pid with a shell command from within dart is hard,
+        // so use a library command, but it's still nice to give the
+        // equivalent command when doing verbose logging.
+        _logging.info('kill $pid');
+        Process.killPid(int.parse(pid));
+      } else {
+        runCheckedSync(['fuser', '-k', '$_serverPort/tcp']);
+      }
+    } catch (e) {}
```

Once we change the earlier part of kill the server, this code will be identical and can probably be factored into a common function.

Also, same question about having both try/catch and runCheckedSync.  Seems like we should just skip the "checked" part of runCheckedSync and then not swallow every exception.